ang/civicase/CaseListTable.js

```diff
@@ -29,13 +30,98 @@
     (function init () {
       bindRouteParamsToScope();
       initiateWatchers();
-
-      $scope.casePlaceholders = $scope.filters.id ? [0] : _.range($scope.page.size);
+      initSubscribers();
       getCases();
+      getAllCasesforSelectAll();
+      $scope.casePlaceholders = $scope.filters.id ? [0] : _.range($scope.page.size);
     }());
 
+    /**
+     * Initiate subscribers
+     */
+    function initSubscribers () {
+      $scope.$on('bulkSelection', function (event, condition) {
+        if (condition === 'none') {
+          deselectAllCases();
+        } else if (condition === 'visible') {
+          selectDisplayedCases();
+        } else if (condition === 'all') {
+          if ($scope.isSelectAllAvailable) {
+            selectEveryCase();
+          }
+        }
+      });
+
+      $scope.$on('bulkSelectionCheckboxClicked', function (event, data) {
+        if (data.selected) {
+          $scope.selectedCases.push(data);
+        } else {
+          _.remove($scope.selectedCases, {
+            id: data.id
+          });
+        }
+      });
+    }
+
+    /**
+     * Asynchronously get all cases for the select all bulk
+     * actions functionality
+     */
+    function getAllCasesforSelectAll () {
+      $scope.selectedCases = []; // Resets all selection.
+      $scope.isSelectAllAvailable = false;
+      var params = getCaseApiParams(angular.extend({}, $scope.filters, $scope.hiddenFilters), $scope.sort, $scope.page);
+      delete params[1];
+      delete params[0][2]['api.Activity.get.1'];
+      params[0][2].return = ['case_type_id', 'status_id', 'is_deleted', 'contacts'];
+      params[0][2].options.limit = 0;
+
+      crmApi(params).then(function (res) {
+        allCases = res[0].values;
+        $scope.isSelectAllAvailable = true;
+      });
+    }
+
+    /**
+     * Select all functionality
+     */
+    function selectEveryCase () {
+      $scope.selectedCases = [];
+      $scope.selectedCases = _.each(allCases, formatCase);
```

Comment:
I though we wont be fomatting all cases, as its a performance problem. Why is it necessary?

Reply:
We need some keys like `item.status` that are created using formatCase Function. I don't think it is a big performance issue as this happens in parallel and is non blocking.